internal/broker/broker.go

```diff
@@ -2,15 +2,738 @@
-type Broker struct{}
+type Broker struct {
+	providerInfo      providers.ProviderInfoer
+	auth              authConfig
+	offlineExpiration time.Duration
+
+	currentSessions        map[string]sessionInfo
+	currentSessionsMu      sync.RWMutex
+	userLastSelectedMode   map[string]string
+	userLastSelectedModeMu sync.Mutex
+	isAuthenticatedCalls   map[string]isAuthenticatedCtx
+	isAuthenticatedCallsMu sync.Mutex
+
+	skipUsernameValidation bool
+
+	privateKey *rsa.PrivateKey
+}
+
+type authConfig struct {
+	cachePath string
+
+	provider    *oidc.Provider
+	providerURL string
+
+	oidcCfg  oidc.Config
+	oauthCfg oauth2.Config
+}
+
+type sessionInfo struct {
+	username string
+	lang     string
+	mode     string
+
+	selectedMode      string
+	firstSelectedMode string
+	allModes          map[string]string
+	attemptsPerMode   map[string]int
+
+	authInfo  map[string]any
+	cachePath string
+
+	currentAuthStep int
+}
+
+type isAuthenticatedCtx struct {
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+type option struct {
+	skipJWTSignatureCheck  bool
+	skipUsernameValidation bool
+	providerInfo           providers.ProviderInfoer
+}
+
+// Option is a func that allows to override some of the broker default settings.
+type Option func(*option)
+
+// New returns a new oidc Broker with the providers listed in the configuration file.
+func New(cfg Config, args ...Option) (b *Broker, err error) {
+	defer decorate.OnError(&err, "could not create broker with provided issuer and client ID")
+
+	opts := option{
+		// This is to avoid too much complexity in the tests.
+		skipJWTSignatureCheck:  false,
+		skipUsernameValidation: false,
+		providerInfo:           providers.CurrentProviderInfo(),
+	}
+	for _, arg := range args {
+		arg(&opts)
+	}
+
+	if cfg.CachePath == "" {
+		return &Broker{}, errors.New("cache path must be provided")
+	}
+
+	clientID := cfg.ClientID
+	issuerURL := cfg.IssuerURL
+	if issuerURL == "" || clientID == "" {
+		return &Broker{}, errors.New("issuer and client ID must be provided")
+	}
+
+	// Generate a new private key for the broker.
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		panic(fmt.Sprintf("could not create an valid rsa key: %v", err))
+	}
+
+	// Create provider
+	provider, err := oidc.NewProvider(context.TODO(), issuerURL)
+	if err != nil {
+		return &Broker{}, err
+	}
+	oidcCfg := oidc.Config{
+		ClientID:                   clientID,
+		InsecureSkipSignatureCheck: opts.skipJWTSignatureCheck,
+	}
+	oauthCfg := oauth2.Config{
+		ClientID: clientID,
+		Endpoint: provider.Endpoint(),
+		Scopes:   append([]string{oidc.ScopeOpenID, "profile", "email"}, opts.providerInfo.AdditionalScopes()...),
+	}
+	authCfg := authConfig{
+		provider:    provider,
+		providerURL: issuerURL,
+		cachePath:   cfg.CachePath,
+		oidcCfg:     oidcCfg,
+		oauthCfg:    oauthCfg,
+	}
+
+	// Set offline expiration
+	var offlineExpiration time.Duration
```

Comment:
Yeah, I tried as much to avoid parsing the configuration in multiple places. The annoying part about it is that parsing some configuration values in the `brokerservice` package wouldn't make sense but not parsing some of the configuration values in that package would also not make sense. So my thinking was:
"generic broker configuration values" (same structure for every authd broker): parsed in `brokerservice`;
"oidc/{provider specific} configuration values": parsed when creating the broker object;